File: tunnel-server/src/main/java/com/tuuzed/tunnel/server/http/HttpServer.java
package com.tuuzed.tunnel.server.http;

import com.tuuzed.tunnel.common.interceptor.HttpRequestInterceptor;
import com.tuuzed.tunnel.server.internal.ServerTunnelSessions;
import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.Channel;
import io.netty.channel.ChannelOption;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.handler.ssl.SslContext;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class HttpServer {
    @NotNull
    private final HttpTunnelRegistry registry;
    @NotNull
    private final ServerBootstrap serverBootstrap;

    public HttpServer(
        @NotNull final NioEventLoopGroup bossGroup,
        @NotNull final NioEventLoopGroup workerGroup,
        @Nullable final SslContext sslContext,
        @NotNull final HttpRequestInterceptor interceptor
    ) {
        this.registry = new HttpTunnelRegistry();
        this.serverBootstrap = new ServerBootstrap();
        this.serverBootstrap.group(bossGroup, workerGroup)
            .channel(NioServerSocketChannel.class)
            .childOption(ChannelOption.AUTO_READ, true)
            .childOption(ChannelOption.SO_KEEPALIVE, true)
            .childHandler(new HttpServerChannelInitializer(sslContext, registry, interceptor));
    }

    @NotNull
    public HttpTunnelRegistry registry() {
        return registry;
    }

    public void serve(@Nullable String bindAddr, int bindPort) throws Exception {
        if (bindAddr != null) {
            serverBootstrap.bind(bindAddr, bindPort).get();
        } else {
            serverBootstrap.bind(bindPort).get();
        }
    }

    @Nullable
    public Channel getSessionChannel(long tunnelToken, long sessionToken) {
        return registry.getSessionChannel(tunnelToken, sessionToken);
    }

    public boolean isRegistered(@NotNull String vhost) {
        return registry.isRegistered(vhost);
    }

    public void register(@NotNull String vhost, @NotNull ServerTunnelSessions tunnelSessions) throws Exception {
        registry.register(vhost, tunnelSessions);
    }

    public void unregister(@Nullable String vhost) {
        registry.unregister(vhost);
    }

    public void destroy() {
        registry.destroy();
    }

}

File: tunnel2-server/src/main/kotlin/tunnel2/server/http/HttpServer.kt
package tunnel2.server.http

import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.ChannelInitializer
import io.netty.channel.ChannelOption
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioServerSocketChannel
import io.netty.handler.codec.http.HttpRequestDecoder
import io.netty.handler.ssl.SslContext
import io.netty.handler.ssl.SslHandler
import tunnel2.common.logging.LoggerFactory
import tunnel2.server.interceptor.HttpRequestInterceptor
import tunnel2.server.tcp.TcpServerChannelHandler

class HttpServer(
    bossGroup: NioEventLoopGroup,
    workerGroup: NioEventLoopGroup,
    private val sslContext: SslContext? = null,
    private val bindAddr: String?,
    private val bindPort: Int,
    private val interceptor: HttpRequestInterceptor
) {
    companion object {
        private val logger = LoggerFactory.getLogger(HttpServer::class.java)
    }

    val registry = HttpRegistry()
    private val serverBootstrap = ServerBootstrap()
    private val https: Boolean get() = sslContext != null

    init {
        this.serverBootstrap.group(bossGroup, workerGroup)
            .channel(NioServerSocketChannel::class.java)
            .childOption(ChannelOption.AUTO_READ, true)
            .childOption(ChannelOption.SO_KEEPALIVE, true)
            .childHandler(createChannelInitializer(sslContext))
    }

    fun start() {
        if (bindAddr != null) {
            serverBootstrap.bind(bindAddr, bindPort).get()
        } else {
            serverBootstrap.bind(bindPort).get()
        }
        logger.info("Serving {} on {} port {}",
            if (https) "https" else "http",
            bindAddr ?: "any address",
            bindPort
        )
    }

    fun shutdown() = registry.destroy()
    fun destroy() = shutdown()

    private fun createChannelInitializer(sslContext: SslContext?) = object : ChannelInitializer<SocketChannel>() {
        override fun initChannel(ch: SocketChannel?) {
            ch ?: return
            if (sslContext != null) {
                ch.pipeline().addFirst(
                    SslHandler(sslContext.newEngine(ch.alloc()))
                )
            }
            ch.pipeline()
                .addLast(HttpRequestDecoder())
                .addLast(HttpServerChannelHandler(registry, interceptor))
        }
    }
}